From 9762c2087524cb6789bd8f7da032dc14a5f2c2c9 Mon Sep 17 00:00:00 2001
From: qua3k <cliffmaceyak@gmail.com>
Date: Tue, 19 Oct 2021 07:36:10 -0400
Subject: [PATCH] Initial support for disabling V8's JIT

This adds a flag in `chrome://flags` for toggling the V8 JIT
compiler; it is default disabled.
---
 chrome/browser/about_flags.cc              | 5 +++++
 chrome/browser/flag-metadata.json          | 7 +++++++
 chrome/browser/flag-never-expire-list.json | 1 +
 chrome/browser/flag_descriptions.cc        | 5 +++++
 chrome/browser/flag_descriptions.h         | 3 +++
 gin/gin_features.cc                        | 3 +++
 gin/gin_features.h                         | 1 +
 gin/v8_initializer.cc                      | 4 ++++
 8 files changed, 29 insertions(+)

diff --git a/chrome/browser/about_flags.cc b/chrome/browser/about_flags.cc
index f2e9d3f0317a8..ada0bd5557036 100644
--- a/chrome/browser/about_flags.cc
+++ b/chrome/browser/about_flags.cc
@@ -172,6 +172,7 @@
 #include "device/gamepad/public/cpp/gamepad_features.h"
 #include "device/vr/buildflags/buildflags.h"
 #include "extensions/buildflags/buildflags.h"
+#include "gin/gin_features.h"
 #include "gpu/config/gpu_finch_features.h"
 #include "gpu/config/gpu_switches.h"
 #include "media/audio/audio_features.h"
@@ -6275,6 +6276,10 @@ const FeatureEntry kFeatureEntries[] = {
      flag_descriptions::kDisableProcessReuseDescription, kOsDesktop,
      FEATURE_VALUE_TYPE(features::kDisableProcessReuse)},
 
+    {"disable-v8-jit", flag_descriptions::kV8NoJITName,
+     flag_descriptions::kV8NoJITDescription, kOsAll,
+     FEATURE_VALUE_TYPE(features::kV8NoJIT)},
+
 #if !BUILDFLAG(IS_ANDROID)
     {"enable-accessibility-live-caption",
      flag_descriptions::kEnableAccessibilityLiveCaptionName,
diff --git a/chrome/browser/flag-metadata.json b/chrome/browser/flag-metadata.json
index 4628d3ef97b78..c295e42fec16b 100644
--- a/chrome/browser/flag-metadata.json
+++ b/chrome/browser/flag-metadata.json
@@ -1431,6 +1431,13 @@
     // issues.
     "expiry_milestone": -1
   },
+  {
+    "name": "disable-v8-jit",
+    "owners": [ "cliffmaceyak@gmail.com" ],
+    // This flag does not expire because it allows users to disable the
+    // V8 Just-In-Time compiler.
+    "expiry_milestone": -1
+  },
   {
     "name": "disable-virtual-keyboard",
     "owners": [ "dvallet", "essential-inputs-team@google.com" ],
diff --git a/chrome/browser/flag-never-expire-list.json b/chrome/browser/flag-never-expire-list.json
index 138ab3676967e..3ea6fd25ad954 100644
--- a/chrome/browser/flag-never-expire-list.json
+++ b/chrome/browser/flag-never-expire-list.json
@@ -30,6 +30,7 @@
   "disable-explicit-dma-fences",
   "disable-javascript-harmony-shipping",
   "disable-threaded-scrolling",
+  "disable-v8-jit",
   "disable-webrtc-hw-decoding",
   "disable-webrtc-hw-encoding",
   "disallow-doc-written-script-loads",
diff --git a/chrome/browser/flag_descriptions.cc b/chrome/browser/flag_descriptions.cc
index 4cad11883511b..efb94b9e092f1 100644
--- a/chrome/browser/flag_descriptions.cc
+++ b/chrome/browser/flag_descriptions.cc
@@ -3055,6 +3055,11 @@ const char kVCPortraitRelightingDescription[] =
     "Enables portrait relighting feature for video conferencing on "
     "chromebooks. THIS WILL OVERRIDE BACKGROUND BLUR & REPLACE.";
 
+const char kV8NoJITName[] = "Disable V8 JIT (JITless)";
+const char kV8NoJITDescription[] =
+    "Security mode that disables all runtime allocation of "
+    "executable memory";
+
 const char kV8VmFutureName[] = "Future V8 VM features";
 const char kV8VmFutureDescription[] =
     "This enables upcoming and experimental V8 VM features. "
diff --git a/chrome/browser/flag_descriptions.h b/chrome/browser/flag_descriptions.h
index e7e92f57e3d98..0789c96e2df67 100644
--- a/chrome/browser/flag_descriptions.h
+++ b/chrome/browser/flag_descriptions.h
@@ -1731,6 +1731,9 @@ extern const char kVCBackgroundReplaceDescription[];
 extern const char kVCPortraitRelightingName[];
 extern const char kVCPortraitRelightingDescription[];
 
+extern const char kV8NoJITName[];
+extern const char kV8NoJITDescription[];
+
 extern const char kV8VmFutureName[];
 extern const char kV8VmFutureDescription[];
 
diff --git a/gin/gin_features.cc b/gin/gin_features.cc
index 035dfdb597614..08c03341593ee 100644
--- a/gin/gin_features.cc
+++ b/gin/gin_features.cc
@@ -69,6 +69,9 @@ BASE_FEATURE(kV8NoReclaimUnmodifiedWrappers,
              "V8NoReclaimUnmodifiedWrappers",
              base::FEATURE_DISABLED_BY_DEFAULT);
 
+// Disables the V8 Just-in-Time compiler
+const base::Feature kV8NoJIT{"V8NoJIT", base::FEATURE_DISABLED_BY_DEFAULT};
+
 // Enables W^X code memory protection in V8.
 // This is enabled in V8 by default. To test the performance impact, we are
 // going to disable this feature in a finch experiment.
diff --git a/gin/gin_features.h b/gin/gin_features.h
index 45f01f71829c4..0170fd951aa67 100644
--- a/gin/gin_features.h
+++ b/gin/gin_features.h
@@ -27,6 +27,7 @@ GIN_EXPORT BASE_DECLARE_FEATURE(kV8FlushEmbeddedBlobICache);
 GIN_EXPORT BASE_DECLARE_FEATURE(kV8LazyFeedbackAllocation);
 GIN_EXPORT BASE_DECLARE_FEATURE(kV8Maglev);
 GIN_EXPORT BASE_DECLARE_FEATURE(kV8NoReclaimUnmodifiedWrappers);
+GIN_EXPORT BASE_DECLARE_FEATURE(kV8NoJIT);
 GIN_EXPORT BASE_DECLARE_FEATURE(kV8CodeMemoryWriteProtection);
 GIN_EXPORT BASE_DECLARE_FEATURE(kV8OffThreadFinalization);
 GIN_EXPORT BASE_DECLARE_FEATURE(kV8OptimizeJavascript);
diff --git a/gin/v8_initializer.cc b/gin/v8_initializer.cc
index 88206801bc1a1..63c65c8f4edc8 100644
--- a/gin/v8_initializer.cc
+++ b/gin/v8_initializer.cc
@@ -299,6 +299,10 @@ void SetFlags(IsolateHolder::ScriptMode mode,
   SetV8FlagsIfOverridden(features::kV8SlowHistograms, "--slow-histograms",
                          "--no-slow-histograms");
 
+  if (base::FeatureList::IsEnabled(features::kV8NoJIT)) {
+    SetV8Flags("--jitless");
+  }
+
   if (base::FeatureList::IsEnabled(features::kV8ConcurrentSparkplug)) {
     if (int max_threads = features::kV8ConcurrentSparkplugMaxThreads.Get()) {
       SetV8FlagsFormatted("--concurrent-sparkplug-max-threads=%i", max_threads);
-- 
2.39.1.windows.1

